File: android/src/main/java/com/unact/yandexmapkit/MapObjectCollectionController.java
package com.unact.yandexmapkit;

import androidx.annotation.NonNull;

import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.map.MapObject;
import com.yandex.mapkit.map.MapObjectCollection;
import com.yandex.mapkit.map.MapObjectTapListener;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MapObjectCollectionController extends MapObjectController implements MapObjectTapListener {
  private final Map<String, MapObjectCollectionController> mapObjectCollections = new HashMap<>();
  private final Map<String, ClusterizedPlacemarkCollectionController> clusterizedPlacemarkCollections =
    new HashMap<>();
  private final Map<String, CircleMapObjectController> circles = new HashMap<>();
  private final Map<String, PlacemarkMapObjectController> placemarks = new HashMap<>();
  private final Map<String, PolygonMapObjectController> polygons = new HashMap<>();
  private final Map<String, PolylineMapObjectController> polylines = new HashMap<>();
  public final MapObjectCollection mapObjectCollection;
  private boolean consumeTapEvents = false;
  private final WeakReference<YandexMapController> controller;
  public final String id;

  public MapObjectCollectionController(
    MapObjectCollection root,
    String id,
    WeakReference<YandexMapController> controller
  ) {
    this.mapObjectCollection = root;
    this.id = id;
    this.controller = controller;

    mapObjectCollection.setUserData(this.id);
    mapObjectCollection.addTapListener(this);
  }

  public MapObjectCollectionController(
    MapObjectCollection parent,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    MapObjectCollection mapObjectCollection = parent.addCollection();

    this.mapObjectCollection = mapObjectCollection;
    this.id = (String) params.get("id");
    this.controller = controller;

    mapObjectCollection.setUserData(this.id);
    mapObjectCollection.addTapListener(this);
    update(params);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void update(Map<String, Object> params) {
    mapObjectCollection.setZIndex(((Double) params.get("zIndex")).floatValue());
    mapObjectCollection.setVisible((Boolean) params.get("isVisible"));
    updateMapObjects((Map<String, Object>) params.get("mapObjects"));

    consumeTapEvents = (Boolean) params.get("consumeTapEvents");
  }

  public void remove() {
    for (ClusterizedPlacemarkCollectionController colController : clusterizedPlacemarkCollections.values()) {
      colController.remove();
    }
    for (CircleMapObjectController circleController : circles.values()) {
      circleController.remove();
    }
    for (MapObjectCollectionController mapObjectCollectionController : mapObjectCollections.values()) {
      mapObjectCollectionController.remove();
    }
    for (PlacemarkMapObjectController placemarkController : placemarks.values()) {
      placemarkController.remove();
    }
    for (PolygonMapObjectController polygonController : polygons.values()) {
      polygonController.remove();
    }
    for (PolylineMapObjectController polylineController : polylines.values()) {
      polylineController.remove();
    }

    clusterizedPlacemarkCollections.clear();
    circles.clear();
    mapObjectCollections.clear();
    placemarks.clear();
    polygons.clear();
    polylines.clear();

    mapObjectCollection.getParent().remove(mapObjectCollection);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  private void updateMapObjects(Map<String, Object> params) {
    addMapObjects((List<Map<String, Object>>) params.get("toAdd"));
    changeMapObjects((List<Map<String, Object>>) params.get("toChange"));
    removeMapObjects((List<Map<String, Object>>) params.get("toRemove"));
  }

  @SuppressWarnings({"ConstantConditions"})
  private void addMapObjects(List<Map<String, Object>> params) {
    for (Map<String, Object> el : params) {
      switch ((String) el.get("type")) {
        case "CircleMapObject":
          addCircle(el);
          break;
        case "MapObjectCollection":
          addMapObjectCollection(el);
          break;
        case "PlacemarkMapObject":
          addPlacemark(el);
          break;
        case "PolygonMapObject":
          addPolygon(el);
          break;
        case "PolylineMapObject":
          addPolyline(el);
          break;
        case "ClusterizedPlacemarkCollection":
          addClusterizedPlacemarkCollection(el);
          break;
        default:
          break;
      }
    }
  }

  @SuppressWarnings({"ConstantConditions"})
  private void changeMapObjects(List<Map<String, Object>> params) {
    for (Map<String, Object> el : params) {
      switch ((String) el.get("type")) {
        case "CircleMapObject":
          changeCircle(el);
          break;
        case "MapObjectCollection":
          changeMapObjectCollection(el);
          break;
        case "PlacemarkMapObject":
          changePlacemark(el);
          break;
        case "PolygonMapObject":
          changePolygon(el);
          break;
        case "PolylineMapObject":
          changePolyline(el);
          break;
        case "ClusterizedPlacemarkCollection":
          changeClusterizedPlacemarkCollection(el);
          break;
        default:
          break;
      }
    }
  }

  @SuppressWarnings({"ConstantConditions"})
  private void removeMapObjects(List<Map<String, Object>> params) {
    for (Map<String, Object> el : params) {
      switch ((String) el.get("type")) {
        case "CircleMapObject":
          removeCircle(el);
          break;
        case "MapObjectCollection":
          removeMapObjectCollection(el);
          break;
        case "PlacemarkMapObject":
          removePlacemark(el);
          break;
        case "PolygonMapObject":
          removePolygon(el);
          break;
        case "PolylineMapObject":
          removePolyline(el);
          break;
        case "ClusterizedPlacemarkCollection":
          removeClusterizedPlacemarkCollection(el);
          break;
        default:
          break;
      }
    }
  }

  private void addCircle(Map<String, Object> params) {
    CircleMapObjectController circleController = new CircleMapObjectController(
      mapObjectCollection,
      params,
      controller
    );

    circles.put(circleController.id, circleController);
  }

  private void changeCircle(Map<String, Object> params) {
    String id = (String) params.get("id");
    CircleMapObjectController circleController = circles.get(id);

    if (circleController != null) circleController.update(params);
  }

  private void removeCircle(Map<String, Object> params) {
    String id = (String) params.get("id");
    CircleMapObjectController circleController = circles.get(id);

    if (circleController != null) circleController.remove();
    circles.remove(id);
  }

  private void addMapObjectCollection(Map<String, Object> params) {
    MapObjectCollectionController mapObjectCollectionController = new MapObjectCollectionController(
      mapObjectCollection,
      params,
      controller
    );

    mapObjectCollections.put(mapObjectCollectionController.id, mapObjectCollectionController);
  }

  private void changeMapObjectCollection(Map<String, Object> params) {
    String id = (String) params.get("id");
    MapObjectCollectionController mapObjectCollectionController = mapObjectCollections.get(id);

    if (mapObjectCollectionController != null) mapObjectCollectionController.update(params);
  }

  private void removeMapObjectCollection(Map<String, Object> params) {
    String id = (String) params.get("id");
    MapObjectCollectionController mapObjectCollectionController = mapObjectCollections.get(id);

    if (mapObjectCollectionController != null) mapObjectCollectionController.remove();
    mapObjectCollections.remove(id);
  }

  private void addPlacemark(Map<String, Object> params) {
    PlacemarkMapObjectController placemarkController = new PlacemarkMapObjectController(
      mapObjectCollection,
      params,
      controller
    );

    placemarks.put(placemarkController.id, placemarkController);
  }

  private void changePlacemark(Map<String, Object> params) {
    String id = (String) params.get("id");
    PlacemarkMapObjectController placemarkController = placemarks.get(id);

    if (placemarkController != null) placemarkController.update(params);
  }

  private void removePlacemark(Map<String, Object> params) {
    String id = (String) params.get("id");
    PlacemarkMapObjectController placemarkController = placemarks.get(id);

    if (placemarkController != null) placemarkController.remove();
    placemarks.remove(id);
  }

  private void addPolygon(Map<String, Object> params) {
    PolygonMapObjectController polygonController = new PolygonMapObjectController(
      mapObjectCollection,
      params,
      controller
    );

    polygons.put(polygonController.id, polygonController);
  }

  private void changePolygon(Map<String, Object> params) {
    String id = (String) params.get("id");
    PolygonMapObjectController polygonController = polygons.get(id);

    if (polygonController != null) polygonController.update(params);
  }

  private void removePolygon(Map<String, Object> params) {
    String id = (String) params.get("id");
    PolygonMapObjectController polygonController = polygons.get(id);

    if (polygonController != null) polygonController.remove();
    polygons.remove(id);
  }

  private void addPolyline(Map<String, Object> params) {
    PolylineMapObjectController polylineController = new PolylineMapObjectController(
      mapObjectCollection,
      params,
      controller
    );

    polylines.put(polylineController.id, polylineController);
  }

  private void changePolyline(Map<String, Object> params) {
    String id = (String) params.get("id");
    PolylineMapObjectController polylineController = polylines.get(id);

    if (polylineController != null) polylineController.update(params);
  }

  private void removePolyline(Map<String, Object> params) {
    String id = (String) params.get("id");
    PolylineMapObjectController polylineController = polylines.get(id);

    if (polylineController != null) polylineController.remove();
    polylines.remove(id);
  }

  private void addClusterizedPlacemarkCollection(Map<String, Object> params) {
    ClusterizedPlacemarkCollectionController colController = new ClusterizedPlacemarkCollectionController(
      mapObjectCollection,
      params,
      controller
    );

    clusterizedPlacemarkCollections.put(colController.id, colController);
  }

  private void changeClusterizedPlacemarkCollection(Map<String, Object> params) {
    String id = (String) params.get("id");
    ClusterizedPlacemarkCollectionController colController = clusterizedPlacemarkCollections.get(id);

    if (colController != null) colController.update(params);
  }

  private void removeClusterizedPlacemarkCollection(Map<String, Object> params) {
    String id = (String) params.get("id");
    ClusterizedPlacemarkCollectionController colController = clusterizedPlacemarkCollections.get(id);

    if (colController != null) colController.remove();
    clusterizedPlacemarkCollections.remove(id);
  }

  @Override
  public boolean onMapObjectTap(@NonNull MapObject mapObject, @NonNull Point point) {
    controller.get().mapObjectTap(id, point);

    return consumeTapEvents;
  }
}

File: ios/Classes/MapObjectCollectionController.swift
import YandexMapsMobile

class MapObjectCollectionController: NSObject, MapObjectController, YMKMapObjectTapListener {
  private var mapObjectCollections: [String: MapObjectCollectionController] = [:]
  private var clusterizedPlacemarkCollections: [String: ClusterizedPlacemarkCollectionController] = [:]
  private var placemarks: [String: PlacemarkMapObjectController] = [:]
  private var circles: [String: CircleMapObjectController] = [:]
  private var polylines: [String: PolylineMapObjectController] = [:]
  private var polygons: [String: PolygonMapObjectController] = [:]
  public let mapObjectCollection: YMKMapObjectCollection
  private var consumeTapEvents: Bool = false
  public weak var controller: YandexMapController?
  public let id: String

  internal init(
    root: YMKMapObjectCollection,
    id: String,
    controller: YandexMapController
  ) {
    self.mapObjectCollection = root
    self.id = id
    self.controller = controller

    super.init()

    mapObjectCollection.userData = self.id
    mapObjectCollection.addTapListener(with: self)
  }

  public required init(
    parent: YMKMapObjectCollection,
    params: [String: Any],
    controller: YandexMapController
  ) {
    let mapObjectCollection = parent.add()

    self.mapObjectCollection = mapObjectCollection
    self.id = params["id"] as! String
    self.controller = controller

    super.init()

    mapObjectCollection.userData = self.id
    mapObjectCollection.addTapListener(with: self)
    update(params)
  }

  public func update(_ params: [String: Any]) {
    mapObjectCollection.zIndex = (params["zIndex"] as! NSNumber).floatValue
    mapObjectCollection.isVisible = (params["isVisible"] as! NSNumber).boolValue
    updateMapObjects(params["mapObjects"] as! [String: Any])

    consumeTapEvents = (params["consumeTapEvents"] as! NSNumber).boolValue
  }

  public func remove() {
    mapObjectCollections.forEach({ $0.value.remove() })
    clusterizedPlacemarkCollections.forEach({ $0.value.remove() })
    placemarks.forEach({ $0.value.remove() })
    circles.forEach({ $0.value.remove() })
    polylines.forEach({ $0.value.remove() })
    polygons.forEach({ $0.value.remove() })

    mapObjectCollections.removeAll()
    clusterizedPlacemarkCollections.removeAll()
    placemarks.removeAll()
    circles.removeAll()
    polylines.removeAll()
    polygons.removeAll()

    mapObjectCollection.parent.remove(with: mapObjectCollection)
  }

  private func updateMapObjects(_ mapObjects: [String: Any]) {
    addMapObjects(mapObjects["toAdd"] as! [[String: Any]])
    changeMapObjects(mapObjects["toChange"] as! [[String: Any]])
    removeMapObjects(mapObjects["toRemove"] as! [[String: Any]])
  }

  private func addMapObjects(_ toAdd: [[String: Any]]) {
    for el in toAdd {
      switch el["type"] as! String {
      case "CircleMapObject":
        addCircle(el)
        break
      case "MapObjectCollection":
        addMapObjectCollection(el)
        break
      case "PlacemarkMapObject":
        addPlacemark(el)
        break
      case "PolygonMapObject":
        addPolygon(el)
        break
      case "PolylineMapObject":
        addPolyline(el)
        break
      case "ClusterizedPlacemarkCollection":
        addClusterizedPlacemarkCollection(el)
        break
      default:
        break
      }
    }
  }

  private func changeMapObjects(_ toChange: [[String: Any]]) {
    for el in toChange {
      switch el["type"] as! String {
      case "CircleMapObject":
        changeCircle(el)
        break
      case "MapObjectCollection":
        changeMapObjectCollection(el)
        break
      case "PlacemarkMapObject":
        changePlacemark(el)
        break
      case "PolygonMapObject":
        changePolygon(el)
        break
      case "PolylineMapObject":
        changePolyline(el)
        break
      case "ClusterizedPlacemarkCollection":
        changeClusterizedPlacemarkCollection(el)
        break
      default:
        break
      }
    }
  }

  private func removeMapObjects(_ toRemove: [[String: Any]]) {
    for el in toRemove {
      switch el["type"] as! String {
      case "CircleMapObject":
        removeCircle(el)
        break
      case "MapObjectCollection":
        removeMapObjectCollection(el)
        break
      case "PlacemarkMapObject":
        removePlacemark(el)
        break
      case "PolygonMapObject":
        removePolygon(el)
        break
      case "PolylineMapObject":
        removePolyline(el)
        break
      case "ClusterizedPlacemarkCollection":
        removeClusterizedPlacemarkCollection(el)
        break
      default:
        break
      }
    }
  }

  private func addMapObjectCollection(_ params: [String: Any]) {
    let mapObjectCollectionController = MapObjectCollectionController(
      parent: mapObjectCollection,
      params: params,
      controller: controller!
    )

    mapObjectCollections[mapObjectCollectionController.id] = mapObjectCollectionController
  }

  private func changeMapObjectCollection(_ params: [String: Any]) {
    let id = params["id"] as! String

    mapObjectCollections[id]?.update(params)
  }

  private func removeMapObjectCollection(_ params: [String: Any]) {
    let id = params["id"] as! String

    mapObjectCollections[id]?.remove()
    mapObjectCollections.removeValue(forKey: id)
  }

  private func addPlacemark(_ params: [String: Any]) {
    let placemarkController = PlacemarkMapObjectController(
      parent: mapObjectCollection,
      params: params,
      controller: controller!
    )

    placemarks[placemarkController.id] = placemarkController
  }

  private func changePlacemark(_ params: [String: Any]) {
    let id = params["id"] as! String

    placemarks[id]?.update(params)
  }

  private func removePlacemark(_ params: [String: Any]) {
    let id = params["id"] as! String

    placemarks[id]?.remove()
    placemarks.removeValue(forKey: id)
  }

  private func addCircle(_ params: [String: Any]) {
    let circleController = CircleMapObjectController(
      parent: mapObjectCollection,
      params: params,
      controller: controller!
    )

    circles[circleController.id] = circleController
  }

  private func changeCircle(_ params: [String: Any]) {
    let id = params["id"] as! String

    circles[id]?.update(params)
  }

  private func removeCircle(_ params: [String: Any]) {
    let id = params["id"] as! String

    circles[id]?.remove()
    circles.removeValue(forKey: id)
  }

  private func addPolyline(_ params: [String: Any]) {
    let polylineController = PolylineMapObjectController(
      parent: mapObjectCollection,
      params: params,
      controller: controller!
    )

    polylines[polylineController.id] = polylineController
  }

  private func changePolyline(_ params: [String: Any]) {
    let id = params["id"] as! String

    polylines[id]?.update(params)
  }

  private func removePolyline(_ params: [String: Any]) {
    let id = params["id"] as! String

    polylines[id]?.remove()
    polylines.removeValue(forKey: id)
  }

  private func addPolygon(_ params: [String: Any]) {
    let polygonController = PolygonMapObjectController(
      parent: mapObjectCollection,
      params: params,
      controller: controller!
    )

    polygons[polygonController.id] = polygonController
  }

  private func changePolygon(_ params: [String: Any]) {
    let id = params["id"] as! String

    polygons[id]?.update(params)
  }

  private func removePolygon(_ params: [String: Any]) {
    let id = params["id"] as! String

    polygons[id]?.remove()
    polygons.removeValue(forKey: id)
  }

  private func addClusterizedPlacemarkCollection(_ params: [String: Any]) {
    let clusterizedPlacemarkCollectionController = ClusterizedPlacemarkCollectionController(
      parent: mapObjectCollection,
      params: params,
      controller: controller!
    )

    clusterizedPlacemarkCollections[clusterizedPlacemarkCollectionController.id] =
      clusterizedPlacemarkCollectionController
  }

  private func changeClusterizedPlacemarkCollection(_ params: [String: Any]) {
    let id = params["id"] as! String

    clusterizedPlacemarkCollections[id]?.update(params)
  }

  private func removeClusterizedPlacemarkCollection(_ params: [String: Any]) {
    let id = params["id"] as! String

    clusterizedPlacemarkCollections[id]?.remove()
    clusterizedPlacemarkCollections.removeValue(forKey: id)
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    controller!.mapObjectTap(id: id, point: point)

    return consumeTapEvents
  }
}
